### ISSUE TITLE
[Task 28] Add Synchronous RLOO Training

### ISSUE THREAD

**Issue by Dayuxiaoshui** [redai-infra/Relax#176]
## Background

Relax currently supports GRPO and several related policy-gradient estimators, but it does not expose REINFORCE Leave-One-Out (RLOO) as a public baseline. RLOO samples multiple responses for the same prompt and uses the other responses in that prompt group as an unbiased, parameter-free baseline. This reduces policy-gradient variance without adding a critic model.

The implementation should follow *Back to Basics: Revisiting REINFORCE Style Optimization for Learning from Human Feedback in LLMs* (ACL 2024, arXiv:2402.14740) and integrate with Relax's existing synchronous colocate training path.

## Objective

Add `rloo` to the algorithm registry and implement its reward baseline, sequence advantage, policy loss, and monitoring metrics for synchronous training only.

For a prompt group with `K` sampled responses and finite scalar task rewards `r_1, ..., r_K`, define:

```text
KL_i = sum_t m_i,t * (log pi_old(y_i,t | ...) - log pi_ref(y_i,t | ...))
R_i = r_i - beta * KL_i
b_i = (sum_j R_j - R_i) / (K - 1)
A_i = R_i - b_i
L_i = -stop_gradient(A_i) * sum_t m_i,t * log pi_theta(y_i,t | x_i, y_i,<t)
```

where `m_i,t` is the response loss mask. The group objective is the mean of the per-sequence losses. The KL-shaped reward is configured through `--kl-coef`; `--use-kl-loss` is rejected for RLOO to avoid applying the KL penalty twice.

## Scope

- Register `rloo` with the same synchronous actor/rollout role topology used by critic-free policy-gradient algorithms.
- Accept `--advantage-estimator rloo` and reject `--fully-async` and hybrid-async modes for this estimator.
- Require `--n-samples-per-prompt >= 2`.
- Compute sequence KL after the old-policy/reference forwards, gather shaped rewards and `Sample.group_index` across the explicit DP group, then compute each leave-one-out baseline globally. Do not rely on samples being contiguous or colocated on one DP rank.
- Broadcast each scalar sequence advantage to the valid response-token shape used by the Megatron loss path.
- Implement the unclipped sequence-level REINFORCE objective. Existing PPO/GRPO/GSPO/SAPO/CISPO loss behavior must remain unchanged.
- Preserve existing KL metrics and expose RLOO's sequence KL/reward-shaping metrics.
- Report at least:
  - `rloo_shaped_reward`
  - `rloo_sequence_kl`
  - `rloo_baseline`
  - `rloo_advantage`
  - `rloo_advantage_abs`
  - `rloo_empty_response_fraction`
  - existing raw reward, loss, policy-gradient loss, and policy KL metrics
- Add a synchronous example launch script and a reproducible GRPO/RLOO comparison report.

### Group size

- `K < 2` is invalid and must raise a readable configuration error before services are launched.
- Every observed `group_index` must contain exactly `--n-samples-per-prompt` samples. Missing or oversized groups must raise `ValueError` and identify the group and observed size.

### Padding and masks

- Prompt tokens and padding tokens never contribute to the RLOO policy loss.
- Masked response tokens do not contribute to the loss or its denominator.
- Static CP shards first reconstruct the full masked sequence KL/log-probability. Only CP rank 0 contributes the replicated sequence loss, with metric scaling that preserves the unsplit value.

### Empty responses

- A response with zero valid response tokens is accepted as rollout data but contributes zero policy loss and zero gradient.
- Its scalar reward still participates in the leave-one-out baselines of the other samples in the prompt group.
- Empty-response frequency is exposed as a monitoring metric.

### Invalid rewards

- `NaN`, `+inf`, and `-inf` task rewards are rejected before baseline computation.
- The error must identify the sample position or prompt group. Invalid values must not be silently clamped or included in metrics.

## Implementation Notes

Likely integration points:

- `relax/core/registry.py`: register the `rloo` role graph.
- `relax/utils/arguments.py`: parser choice and synchronous/group-size validation.
- `relax/utils/utils.py`: carry stable prompt-group IDs and validate rollout rewards/groups.
- `relax/utils/training/ppo_utils.py`: pure baseline/advantage and element-wise loss helpers.
- `relax/backends/megatron/loss.py`: CP sequence aggregation, explicit DP gather, advantage broadcast, loss, and metrics.
- `relax/utils/training/data_fields.py`: carry RLOO monitoring fields through TransferQueue.
- `tests/`: pure formula tests plus integration-focused mask, DP, and CP tests.
- `examples/algorithms/`: synchronous RLOO launch script.

The independent `Advantages` Ray Serve deployment is not part of this task because fully asynchronous RLOO is explicitly out of scope.

### Formula tests

Use hand-constructed prompt groups with non-contiguous `group_index` values and compare every element against an independently written reference:

```text
group A rewards: [1.0, 0.0, 0.5]
baseline:        [0.25, 0.75, 0.50]
advantage:       [0.75, -0.75, 0.00]
```

Verify the policy loss from explicit token log-probabilities and response masks, including mixed response lengths and padded tokens.

### Boundary tests

- `K = 1` configuration rejection.
- Incomplete and oversized prompt groups.
- Zero-length response and all-zero response mask.
- `NaN`, `+inf`, and `-inf` rewards.
- Float32 and float64 formula consistency.

### Distributed invariance tests

- Split prompt groups across two DP ranks, gather shaped rewards/group IDs with the explicit DP process group, and verify the resulting local baselines/advantages match the unpartitioned reference.
- Split response tokens into simulated CP shards with padding/masks, reduce the shard losses, and verify equality with the unsplit loss on valid tokens.
- Verify sequence-count normalization. Static CP may require the Megatron Bridge `--calculate-per-token-loss` compatibility flag, but RLOO must still reduce gradients and metrics by sequence count rather than valid-token count.

### Regression tests

- Existing default `grpo` parser behavior and role graph are unchanged.
- Existing GRPO return, policy loss, KL, and reward-normalization tests continue to pass.
- `rloo` appears in the registry and resolves to the synchronous colocate role set.
- `rloo --fully-async` fails fast with a readable message.

## Two-GPU Experiment

Run matched experiments on one host with 2 x RTX 5090 (32 GB each), using Qwen3-0.6B and GSM8K to keep the validation affordable.

Hold constant across both runs:

- initial checkpoint and reference checkpoint
- dataset and prompt order/seed
- `rollout_batch_size`, `n_samples_per_prompt`, and `global_batch_size`
- rollout temperature and maximum response length
- optimizer, learning rate, KL coefficient, and number of rollout steps
- actor/rollout placement and SGLang memory fraction

Change only `--advantage-estimator {grpo,rloo}` and algorithm-specific reward normalization. Use at least 10 completed train steps for a smoke comparison; prefer 50-100 steps for a directional training comparison.

Report the same step window for both runs:

| Category | Metrics |
| --- | --- |
| Reward | mean `rollout/raw_reward`, final-window mean, optional eval pass@1 |
| Loss | mean and standard deviation of `train/pg_loss` and `train/loss` |
| KL | mean policy KL for both runs, plus RLOO sequence KL and GRPO explicit KL loss |
| Throughput | samples/s, response tokens/s, actor train step time, rollout time, end-to-end wall time |
| Stability | grad norm, non-finite metric count, OOM/restart count |

The report must state that a short two-GPU run validates execution and metric wiring but is not sufficient to establish algorithmic superiority.

## Acceptance Criteria

- Hand-constructed batches produce element-wise baselines, advantages, and losses matching the formulas above.
- Group size, incomplete groups, padding, empty responses, and non-finite rewards have the explicit behavior documented above.
- DP and CP partitioning do not change valid-token statistics or the reduced loss.
- A small-model synchronous RLOO training run completes on 2 x RTX 5090 and is compared with a matched GRPO run for reward, loss, KL, and throughput.
- The default algorithm remains `grpo`, and existing algorithms' behavior is unchanged.
- Fully asynchronous RLOO is rejected and remains out of scope.
- Focused tests pass and `pre-commit run --all-files` is clean, excluding hardware-dependent multi-node integration tests with an explicit reason.

## Non-Goals

- Fully asynchronous or hybrid-async RLOO.
- A critic/value model.
- New dependencies.
- Changes to Controller, Service, or launcher behavior beyond selecting the registered synchronous role graph.
- Claims of statistically significant RLOO-versus-GRPO quality from a short smoke run.

## References

- Ahmadian et al., *Back to Basics: Revisiting REINFORCE Style Optimization for Learning from Human Feedback in LLMs*, ACL 2024: https://arxiv.org/abs/2402.14740
- Hugging Face TRL RLOO documentation: https://huggingface.co/docs/trl/rloo_trainer
